src/cargo/core/gc.rs

```diff
@@ -0,0 +1,550 @@
+    ///
+    /// * `kinds` is a list of [`AutoGcKind`] that is being requested to
+    ///   perform. This corresponds to the `cargo clean --gc` flag. If empty,
+    ///   no config options are incorporated.
+    /// * `max_download_age` is the `--max-download-age` CLI option which
+    ///   requires special handling since it implicitly overlaps two options.
+    ///   It will use the newer value of either this or the explicit value.
+    ///
+    /// The `kinds` list is used in a few different ways:
+    ///
+    /// * If empty, uses only the options the user specified on the
+    ///   command-line, like `cargo clean --max-crate-size=…`.
+    /// * If the user specified a `cargo clean --gc` option, then the `kinds`
+    ///   list is filled in with whatever `--gc` option the user picked, and
+    ///   then this function *merges* the settings between the requested
+    ///   `--gc` option and any options that were explicitly specified.
+    /// * [`AutoGcKind::All`] is used in `cargo clean` when no options are
+    ///   specified.
+    pub fn update_for_auto_gc(
+        &mut self,
+        config: &Config,
+        kinds: &[AutoGcKind],
+        max_download_age: Option<Duration>,
+    ) -> CargoResult<()> {
+        let auto_config = config
+            .get::<Option<AutoConfig>>("gc.auto")?
+            .unwrap_or_default();
+        self.update_for_auto_gc_config(&auto_config, kinds, max_download_age)
+    }
+
+    fn update_for_auto_gc_config(
+        &mut self,
+        auto_config: &AutoConfig,
+        kinds: &[AutoGcKind],
+        max_download_age: Option<Duration>,
+    ) -> CargoResult<()> {
+        for kind in kinds {
+            if matches!(kind, AutoGcKind::All | AutoGcKind::Download) {
+                self.max_src_age = newer_time_span_for_config(
+                    self.max_src_age,
+                    "gc.auto.max-src-age",
+                    auto_config.max_src_age.as_deref().unwrap_or("1 month"),
+                )?;
+                self.max_crate_age = newer_time_span_for_config(
+                    self.max_crate_age,
+                    "gc.auto.max-crate-age",
+                    auto_config.max_crate_age.as_deref().unwrap_or("3 months"),
+                )?;
+                self.max_index_age = newer_time_span_for_config(
+                    self.max_index_age,
+                    "gc.auto.max-index-age",
+                    auto_config.max_index_age.as_deref().unwrap_or("3 months"),
+                )?;
+                self.max_git_co_age = newer_time_span_for_config(
+                    self.max_git_co_age,
+                    "gc.auto.max-git-co-age",
+                    auto_config.max_git_co_age.as_deref().unwrap_or("1 month"),
+                )?;
+                self.max_git_db_age = newer_time_span_for_config(
+                    self.max_git_db_age,
+                    "gc.auto.max-git-db-age",
+                    auto_config.max_git_db_age.as_deref().unwrap_or("3 months"),
```

Comment:
Should we use constants for the defaults.

Looks like they are divided between being a primary and secondary source

Reply:
I pulled these into constants, though I'm not sure I feel entirely comfortable with that. My thinking is that we may want to tweak these defaults in ways that are unique to each setting (like maybe git db should be longer than `.crate`, since it can be much more expensive to download). But I also don't really know if that will actually happen, and the structure can be tweaked if necessary.